### PR TITLE
8221988: add possibility to build with Visual Studio 2019

### PR DESCRIPTION
**Pull request by apavlyutkin**
Hi!

Please review backport of 8221988. This one opens a bunch of backports enabling support of VS2019. **WARNING**: the patch DOES NOT enable the building with MSVS2019 toolkit, that requires 5-6 more backports. The patch ONLY adds recognition of MSVS2019 to the build system.

The 11u patch applied with the following changes:

- changes to absent documentation files skipped
- resolved baseline conflicts toolchain_windows.m4
- valid MSVS versions were ascending sorted

Verification: `bash configure` under VS2019 prompt.

Regression:

- VS2017 full build (both 64/32-bit): ok
- VS2015 full build: ok (requires --with-extra-cflags='/EHsc /wd4091' --with-extra-cxxflags='/EHsc /wd4091')
- VS2013 full build: ok
- VS2012 full build: ok
- VS2010 full build: ok

@RealCLanger @mdoerr if you take a look at this I will be very grateful

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8221988](https://bugs.openjdk.java.net/browse/JDK-8221988): add possibility to build with Visual Studio 2019

### Reviewers
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.java.net/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk8u-dev pull/33/head:pull/33` \
`$ git checkout pull/33`

Update a local copy of the PR: \
`$ git checkout pull/33` \
`$ git pull https://git.openjdk.java.net/jdk8u-dev pull/33/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 33`

View PR using the GUI difftool: \
`$ git pr show -t 33`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk8u-dev/pull/33.diff">https://git.openjdk.java.net/jdk8u-dev/pull/33.diff</a>

</details>

### CODE REVIEW

File: common/autoconf/toolchain_windows.m4 [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 ################################################################################
 # The order of these defines the priority by which we try to find them.
-VALID_VS_VERSIONS="2010 2012 2013 2015 2017"
+VALID_VS_VERSIONS="2017 2019 2013 2015 2012 2010"
```
Comment (RealCLanger) [inline common/autoconf/toolchain_windows.m4:28]: I'm not sure whether it's appropriate to put 2017 and 2019 in front of the list - maybe for compatibilities sake we should add 2019 at the end?

**Comment by RealCLanger** [comment]
Hi @apavlyutkin,
apart from my other comment I think the backport looks okay. However, I've no means to test it. Probably somebody more involved with OpenJDK 8 (and its Windows build) should have a look...

File: common/autoconf/toolchain_windows.m4 (hunk 1 above)
Comment (apavlyutkin) [reply common/autoconf/toolchain_windows.m4:28]: I completely agree with you but that is how it comes from 11u. I'm not sure if it's a good idea to make such changes as part of the backport. On my last backport to 8u @gnu-andrew made me to eliminate every signle whitespace of unwanted delta.

Comment (RealCLanger) [reply common/autoconf/toolchain_windows.m4:28]: Yes, let @gnu-andrew or @jerboaa deliver the judgement here.

Comment (jerboaa) [reply common/autoconf/toolchain_windows.m4:28]: My understanding is that the order for different JDK releases may differ. Newer JDKs might want to build with the newest toolchain if possible. That's different for update releases. So if we backport this, versions for 2017 and 2019 should go to the end. See for example `8196108: Add build support for VS 2015/2017 ` which added this 4 years ago for 2015 and 2017.

TLDR; For compatibility we shall append.

File: common/autoconf/toolchain_windows.m4 [hunk 2]
```diff
@@ -25,7 +25,7 @@
 
 ################################################################################
 # The order of these defines the priority by which we try to find them.
-VALID_VS_VERSIONS="2010 2012 2013 2015 2017"
+VALID_VS_VERSIONS="2013 2015 2012 2010 2017 2019"
```
Comment (jerboaa) [inline common/autoconf/toolchain_windows.m4:28]: Order should be: `2010 2012 2013 2015 2017 2019`

**Review by jerboaa — COMMENTED** [review]
With which versions of Visual Studio has this been tested?

File: common/autoconf/toolchain_windows.m4 (hunk 2 above)
Comment (apavlyutkin) [reply common/autoconf/toolchain_windows.m4]: 2017 (build) and 2019 (configuration). To me it's enough at this stage, the changes do not touch older toolkits, only define new 2019. Moreover it's a bit difficult to test for now, because 5-6 more backports required to enable VS2019, and few of them require complete regression. For now VS2019 build will fail immediately

Comment (apavlyutkin) [reply common/autoconf/toolchain_windows.m4]: Fixed. The patch tested for all 2010-2019 toolkits, the PR header was updated accordingly

Comment (apavlyutkin) [reply common/autoconf/toolchain_windows.m4]: Ping

**Comment by karianna** [comment]
Hey all - missed this whilst travelling.  Happy to help test this internally at MSFT

**Comment by apavlyutkin** [comment]
Ping. Can we proceed somehow? There are several more backports required to enable MSVC 2019 support, but they all depend on this one. Thank you

**Review by jerboaa — APPROVED** [review]
OK, though, I have no way to test this myself.

**Comment by apavlyutkin** [comment]
[JDK-8221988](https://bugs.openjdk.java.net/browse/JDK-8221988) labelled with jdk8u-fix-request

**Comment by gnu-andrew** [comment]
> With which versions of Visual Studio has this been tested?

GHA is testing VS2010 & VS2017, for what it's worth. Currently wondering what broke MacOS (not specific to this fix)

**Comment by gnu-andrew** [comment]
> Hey all - missed this whilst travelling. Happy to help test this internally at MSFT

Extra testing is always welcome :)

**Review by gnu-andrew — APPROVED** [review]
Backport looks good to me and GHA suggests it doesn't break existing builds on VS2010 & VS2017.

There is documentation in 8u but it is in `README-builds.html` and currently only seems to mention VS2010. We should probably do an update of the documentation, starting with a JDK-8176509 backport so it reflects the current status quo, especially after the move to Git. It's fine to leave this change out here though.

**Comment by karianna** [comment]
Seemed OK with our internal testing FWIW - sorry to be late to the party
here!

Cheers,
Martijn
